astropy/coordinates/baseframe.py

```diff
@@ -424,39 +424,121 @@ def __init__(self, *args, copy=True, representation_type=None,
         representation_type = kwargs.pop('representation_type', representation_type)
 
         if representation_type is not None or differential_type is not None:
+            representation_type, differential_type = self._infer_representation_differential_type(
+                representation_type, differential_type)
+            self.set_representation_cls(representation_type, **differential_type)
+        else:
+            self._representation = {'base': self.default_representation,
+                                    's': self.default_differential}
 
-            if representation_type is None:
-                representation_type = self.default_representation
+        representation_data, differential_data = self._infer_representation_differential_data(
+            args, copy, kwargs)
+
+        if differential_data:
```

Comment:
I think just
```
if differential_data:
    representation_data = representation_data.with_differentials({'s': differential_data})
self._data = representation_data
```
would be slightly clearer. If so, I think one can also consider moving the `if` statement inside the function call (which could then just be `_infer_representation_data(...)`) - in the end the differentials are supposed to be attached to a representation.

Reply:
Nice! I like this solution much more. Changing.